File: solutions/elixir/protein-translation/1/lib/protein_translation.ex
defmodule ProteinTranslation do
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {:ok, list(String.t())} | {:error, String.t()}
  def of_rna(""), do: {:ok, []}
  def of_rna(rna) do
    Regex.scan(~r/.{3}/, rna)
    |> List.flatten()
    |> Enum.reduce_while([], fn c, acc ->
      case of_codon(c) do
        {:error, err} -> {:halt, err}
        {:ok, "STOP"} -> {:halt, acc ++ ["STOP"]}
        {:ok, protein} ->
          if List.ends_with?(acc, ["STOP"]) do
            {:cont, acc}
          else
            {:cont, acc ++ [protein]}
          end
      end
    end)
    |> case do
      err when is_binary(err) or err === [] -> {:error, "invalid RNA"}
      proteins when is_list(proteins) ->
        if rem(String.length(rna), 3) === 0 or List.ends_with?(proteins, ["STOP"]) do
          {:ok, Enum.filter(proteins, & &1 !== "STOP")}
        else
          {:error, "invalid RNA"}
        end
    end
  end

  @doc """
  Given a codon, return the corresponding protein
  """
  @spec of_codon(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def of_codon("AUG"), do: {:ok, "Methionine"}
  def of_codon("UGG"), do: {:ok, "Tryptophan"}
  def of_codon(c) when c in ["UUA", "UUG"], do: {:ok, "Leucine"}
  def of_codon(c) when c in ["UGU", "UGC"], do: {:ok, "Cysteine"}
  def of_codon(c) when c in ["UAU", "UAC"], do: {:ok, "Tyrosine"}
  def of_codon(c) when c in ["UUU", "UUC"], do: {:ok, "Phenylalanine"}
  def of_codon(c) when c in ["UCU", "UCC", "UCA", "UCG"], do: {:ok, "Serine"}
  def of_codon(c) when c in ["UAA", "UAG", "UGA"], do: {:ok, "STOP"}
  def of_codon(_), do: {:error, "invalid codon"}
end

File: solutions/typescript/protein-translation/1/protein-translation.ts
const CODONS_REGEXP = /.{1,3}/g

export type Codon = 'AUG' | 'UUU' | 'UUC' | 'UUA' | 'UUG' | 'UCU' | 'UCC' | 'UCA' | 'UCG' | 'UAU' | 'UAC' | 'UGU' | 'UGC' | 'UGG' | 'UAA' | 'UAG' | 'UGA'
export type AminoAcid = 'Methionine' | 'Phenylalanine' | 'Leucine' | 'Serine' | 'Tyrosine' | 'Cysteine' | 'Tryptophan'
export type Protein = AminoAcid[]

/**
 * Translates an RNA sequence into a chain of amino acids which forms a protein.
 *
 * @param rna - The RNA sequence to translate
 * @returns Array of amino acids names (e.g., ['Methionine', 'Phenylalanine'])
 * @throws {Error} If RNA contains invalid codons or is incomplete
 *
 * @example
 * translate('AUGUUUUGG') // ['Methionine', 'Phenylalanine', 'Tryptophan']
 * translate('AUGUAAUUU') // ['Methionine'] (stops at UAA)
 */
export function translate(rna: string): Protein {
  const protein: Protein = []

  for (const codon of rna.matchAll(CODONS_REGEXP)) {
    const aminoAcid: AminoAcid | null = aminoAcidFromCodon(codon[0])
    if (aminoAcid === null) break
    protein.push(aminoAcid)
  }

  return protein
}

/**
 * Translates an RNA codon to its corresponding amino acid.
 *
 * Stop codons (UAA, UAG, UGA) return `null` to signal translation termination.
 *
 * @param codon - A string that should be a valid RNA codon triplet (e.g., 'AUG', 'UUU').
 * @returns The amino acid name (e.g., 'Methionine'), or `null` for stop codons.
 * @throws {Error} If the codon is invalid.
 */
function aminoAcidFromCodon(codon: string): AminoAcid | null {
  switch (codon as Codon) {
    case 'AUG': return 'Methionine'
    case 'UUU':
    case 'UUC': return 'Phenylalanine'
    case 'UUA':
    case 'UUG': return 'Leucine'
    case 'UCU':
    case 'UCC':
    case 'UCA':
    case 'UCG': return 'Serine'
    case 'UAU':
    case 'UAC': return 'Tyrosine'
    case 'UGU':
    case 'UGC': return 'Cysteine'
    case 'UGG': return 'Tryptophan'
    case 'UAA':
    case 'UAG':
    case 'UGA': return null
    default: throw new Error("Invalid codon")
  }
}
